File: tester.c

#include <errno.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "escapeless.h"

static void verror(const char *format, va_list args) {
    fprintf(stderr, "error: ");
    vfprintf(stderr, format, args);
    fprintf(stderr, "\n");
    exit(EXIT_FAILURE);
}

static void error(const char *format, ...) {
    va_list args;
    va_start(args, format);
    verror(format, args);
    va_end(args);
}

static int read_byte(void) {
    int c = getchar();
    if(c == EOF && ferror(stdin))
        error("cannot read input data: %s", strerror(errno));
    return c;
}

static void write_byte(int c) {
    if(putchar(c) == EOF)
        error("cannot write output data: %s", strerror(errno));
}

static void encode(const unsigned char *takeouts, unsigned takeouts_size) {
    unsigned char block[0x100], takeouts_map[0x100];
    unsigned max_block_size, block_size;
    unsigned i;
    int c;

    /* TODO: Validate takeouts size. */

    max_block_size = 0x100 - takeouts_size;

    for(;;) {
        /* Read subsequent block of data. */
        block_size = 0;
        while(block_size < max_block_size && (c = read_byte()) != EOF)
            block[block_size++] = (unsigned char) c;

        if(block_size == 0)
            break;

        /* Encode that block. */
        escapeless_encode(takeouts_map, takeouts, takeouts_size,
                          block, block_size);

        /* Write the substitutes for the takeout bytes. */
        for(i = 0; i != takeouts_size; ++i)
            write_byte(takeouts_map[i]);

        /* Write the block with replaced takeout bytes. */
        for(i = 0; i != block_size; ++i)
            write_byte(block[i]);
    }
}

static void decode(const unsigned char *takeouts, unsigned takeouts_size) {
    unsigned char block[0x100], takeouts_map[0x100];
    unsigned max_block_size, block_size;
    unsigned i;
    int c;

    /* TODO: Validate takeouts size. */

    max_block_size = 0x100 - takeouts_size;

    for(;;) {
        /* Read the substitutes for the takeout bytes. */
        for(i = 0; i < takeouts_size; i++) {
            if((c = read_byte()) == EOF)
                break;
            takeouts_map[i] = (unsigned char) c;
        }

        /* See of we reached the end of file. */
        if(i == 0)
            break;

        if(i < takeouts_size)
            error("unexpected end of file");

        /* Read the payload part of the block. */
        block_size = 0;
        while(block_size < max_block_size && (c = getchar()) != EOF)
            block[block_size++] = (unsigned char) c;

        /* Empty block should have never been encoded. */
        if(block_size == 0)
            error("empty block in input");

        /* Decode the block. */
        escapeless_decode(takeouts_map, takeouts, takeouts_size,
                          block, block_size);

        /* Write decoded block. */
        for(i = 0; i < block_size; i++)
            write_byte(block[i]);
    }
}

static void usage(void) {
    fprintf(stderr, "usage: tester {encode|decode} <takeouts>\n");
    exit(EXIT_FAILURE);
}

int main(int argc, char *argv[]) {
    const char *command, *takeouts;
    size_t takeouts_size;

    if(argc != 3)
        usage();

    takeouts = argv[2];
    takeouts_size = strlen(takeouts);
    if(takeouts_size >= 0x100)
        error("too many takeout bytes");

    command = argv[1];
    if(strcmp(command, "encode") == 0) {
        encode((const unsigned char*) takeouts, (unsigned) takeouts_size);
    } else if(strcmp(command, "decode") == 0) {
        decode((const unsigned char*) takeouts, (unsigned) takeouts_size);
    } else {
        error("unknown command: %s", command);
    }

    return EXIT_SUCCESS;
}

File: python/tester.py
#!/usr/bin/env python

from builtins import bytes
import sys

import escapeless


def error(msg):
    sys.exit('tester: %s' % msg)


def read_bytes():
    while True:
        line = sys.stdin.readline()
        if line == '':
            error('unexpected end of file')

        if line == '\n' or line.startswith('#'):
            continue

        line = line.rstrip()
        break

    b = []
    for group in line.split():
        group = group.split('..')
        if len(group) == 1:
            b.append(int(group[0], 16))
            continue

        begin, end = int(group[0], 16), int(group[1], 16)
        i = begin
        b.append(i)
        while True:
            i = (i + 1) % 0x100
            b.append(i)

            if i == end:
                break

    return bytes(b)


def check_end_of_file():
    while True:
        line = sys.stdin.readline()
        if line == '':
            break

        if line == '\n' or line.startswith('#'):
            continue

        line = line.rstrip()
        error('extra line: %r' % line)


def test_encoding(input, takeouts, encoded):
    block, takeouts_map = escapeless.encode(input, takeouts)

    output = takeouts_map + block
    if output != encoded:
        print('input', input)
        print('takeouts', takeouts)

        print('takeouts_map', takeouts_map)
        print('block', block)
        print('output', output)

        print('encoded', encoded)

        error("encoded block doesn't match")


def test_decoding(input, takeouts, encoded):
    takeouts_size = len(takeouts)
    takeouts_map = encoded[:takeouts_size]
    block = encoded[takeouts_size:]

    block = escapeless.decode(block, takeouts, takeouts_map)

    if block != input:
        print('input', input)
        print('takeouts', takeouts)

        print('takeouts_map', takeouts_map)
        print('block', block)

        error("decoded block doesn't match")


def test(input, takeouts, encoded):
    test_encoding(input, takeouts, encoded)
    test_decoding(input, takeouts, encoded)


def main():
    input = read_bytes()
    takeouts = read_bytes()
    encoded = read_bytes()
    check_end_of_file()

    test(input, takeouts, encoded)


if __name__ == "__main__":
    main()
